working/4200-metaobjects/feature-specification.md

```diff
@@ -0,0 +1,675 @@
+}
+
+class Bird extends Animal static extends AnimalStatics {
+  static bool get flies => true;
+  static bool get walks => true;
+}
+
+class Mammal extends Animal static extends AnimalStatics {
+  static bool get walks => true;
+  static bool get swims => true;
+}
+
+abstract class AnimalStatics {
+  const AnimalStatics(); // Must allow const-able subclasses.
+  bool get swims => false;
+  bool get flies => false;
+  bool get walks => false;
+}
+
+void showCapabilities<X extends Animal static extends AnimalStatics>(X x) {
+  var capabilities = [
+    if (X.walks) "walk",
+    if (X.swims) "swim",
+    if (X.flies) "fly",
+  ];
+  print("$x can do the following: $capabilities");
+}
+
+void main() {
+  showCapabilities(Mammal());
+}
+```
+
+In this case the `static extends` feature is used to provide a default
+implementation of a set of static members (`swims`, `flies`, `walks`).
+The metaobject classes for the subclasses of `Animal` can declare any
+subset of these static members if it wants to override their behavior, and
+the rest are inherited from `AnimalStatics`.
+
+`Mammal.walks` can be invoked according to today's rules about static
+members (nothing new here). However, `Mammal.flies` can be invoked because
+this means `(Mammal).flies`. In other words, even though the invocation
+includes a metaobject, clients can consider `Mammal` to be a class that has
+all of these static members even though only some of them are actually
+declared as static members in `Mammal`. The rest are "inherited" from the
+metaobject class.
+
+In general, `static extends` offers developers greater expressive power
+than `static implements` because it is possible for the metaobject to
+inherit code that developers have written to do whatever they want. A
+metaobject class which was induced by a `static implements` clause, on the
+other hand, will only have methods whose implementation is a forwarding
+invocation of a static member or constructor of the underlying type.
+
+Here is an example where the metaobject is used to provide access to the
+actual type arguments of a given object or type:
+
+```dart
+abstract class CallWithTypeParameters {
+  const CallWithTypeParameters();
+  int get numberOfTypeParameters;
+  R callWithTypeParameter<R>(int number, R callback<X>());
+}
+
+class _EStaticHelper<X, Y> implements CallWithTypeParameters {
+  const _EStaticHelper();
+  int get numberOfTypeParameters => 2;
+  R callWithTypeParameter<R>(int number, R callback<Z>()) {
+    return switch (number) {
+      1 => callback<X>(),
+      2 => callback<Y>(),
+      _ => throw ArgumentError("Expected 1 or 2, got $number."),
+    };
+  }
+}
+
+class E<X, Y> static extends _EStaticHelper<X, Y> {
+  const E();
+  void foo(X x, Y y) {}
+}
+
+void main() {
+  final E<Object, Object> e = E<String, int>();
+
+  // When we don't know the precise type arguments we can't call
+  // `e.foo` safely. But `CallWithTypeParameters` can help!
+  final Object? eType = e.runtimeType;
```

Comment:
May be `final Type eType = e.runtimeType;`? `eType` is treated as non-nullable type below.

Reply:
Ah, I forgot the promotion. We need to use a supertype of `CallWithTypeParameters` as the declared type in order to be able to promote to that type. I added the `if` statement that provides the promotion. Also, I changed the type to `Object` (both `Object?` and `Object` will work, but the latter avoids the detour around null which seems more economical and straightforward ;-).